conans/cli/command.py

```diff
@@ -0,0 +1,282 @@
+    def commands(self):
+        if self._commands is None:
+            self._commands = {}
+            conan_commands_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                                               "commands")
+            for module in pkgutil.iter_modules([conan_commands_path]):
+                self._add_command("conans.cli.commands.{}".format(module.name), module.name)
+            user_commands_path = os.path.join(self._conan.cache_folder, "commands")
+            sys.path.append(user_commands_path)
+            for module in pkgutil.iter_modules([user_commands_path]):
+                if module.name.startswith("cmd_"):
+                    self._add_command(module.name, module.name.replace("cmd_", ""))
+        return self._commands
+
+    @property
+    def groups(self):
+        return self._groups
+
+    def _print_similar(self, command):
+        """ Looks for similar commands and prints them if found.
+        """
+        matches = get_close_matches(
+            word=command, possibilities=self.commands.keys(), n=5, cutoff=0.75)
+
+        if len(matches) == 0:
+            return
+
+        if len(matches) > 1:
+            self._out.writeln("The most similar commands are")
+        else:
+            self._out.writeln("The most similar command is")
+
+        for match in matches:
+            self._out.writeln("    %s" % match)
+
+        self._out.writeln("")
+
+    def help_message(self):
+        self.commands["help"].method(self.conan_api, self.commands, self.groups)
+
+    def run(self, *args):
+        """ Entry point for executing commands, dispatcher to class
+        methods
+        """
+        version = sys.version_info
+        if version.major == 2 or version.minor <= 4:
+            raise ConanException("Unsupported Python version")
+
+        ret_code = SUCCESS
+        try:
+            try:
+                command = args[0][0]
+            except IndexError:  # No parameters
+                self.help_message()
+                return False
+            try:
+                method = self.commands[command].method
+            except KeyError as exc:
+                if command in ["-v", "--version"]:
+                    self._out.success("Conan version %s" % client_version)
+                    return False
+
+                if command in ["-h", "--help"]:
+                    self.help_message()
+                    return False
+
+                self._out.writeln("'%s' is not a Conan command. See 'conan --help'." % command)
+                self._out.writeln("")
+                self._print_similar(command)
+                raise ConanException("Unknown command %s" % str(exc))
+
+            method(args[0][1:], conan_api=self.conan_api, parser=self.commands[command].parser,
+                   commands=self.commands, groups=self.groups) if command == "help" else method(
+                args[0][1:], conan_api=self.conan_api, parser=self.commands[command].parser)
```

Comment:
I'll try to avoid the `if` here. We can pass `commands` and `groups` to every command, they will be ignored in the `**kwargs`